### ISSUE TITLE
Any thoughts on realtime sync backend?

### ISSUE THREAD

**Issue by davej** [lostpebble/pullstate#6]
Hiya, I'm really loving this lib.

I'm switching my app over to Firebase from RESTful APIs. I just wanted to ask if you had any specific thoughts on how this might work with pullstate?

**Comment by lostpebble** [comment]
Hi Dave! Glad to hear you're still enjoying it, was wondering how things were going :)

Interesting use case. I use Firebase in some of my projects as well, but haven't currently with Pullstate (legacy projects that need to be updated).

Are you using Firestore or the Realtime Databse?

In any case, I'm sure they're both similar in their realtime usage. Is it something like this?:

```js
db.collection("cities").doc("SF")
    .onSnapshot(function(doc) {
        console.log(" data: ", doc.data());
    });
```

I think it kinda depends on what you wanna do. In my perspective I'd probably want to update a pullstate store on each update - which in turn will be listened to in my components which are subscribed to that value.

So maybe something like:

```js
function startWatchingCity(cityCode: string) {
  db.collection("cities")
    .doc(cityCode)
    .onSnapshot(function(doc) {
      console.log(" data: ", doc.data());

      CityStore.update(s => {
        s.watchedCities[cityCode] = { updated: new Date(), data: doc.data() };
      });
    });
}
```

This is my incredibly naive, quick code though.. I think it depends on your use case.

Another thing I've added recently to pullstate is "reactions" - although they are undocumented - besides a [small note in the changelog](https://github.com/lostpebble/pullstate/blob/master/Changelog.md#060). In this case (although this is untested, but I feel like it should work) you might be able to do something like:

```js
let currentCity = null;
let currentCityUnsubscribe = () => null;

CityStore.createReaction(s => s.currentCity, (state, rawState, watchedValue) => {
  if (watchedValue !== currentCity) {
    if (currentCity !== null) {
      currentCityUnsubscribe();
    }
  }

  currentCityUnsubscribe = startWatchingCity(watchedValue);
});

function startWatchingCity(cityCode: string) {
  return db.collection("cities")
    .doc(cityCode)
    .onSnapshot(function(doc) {
      console.log(" data: ", doc.data());

      CityStore.update(s => {
        s.watchedCities[cityCode] = { updated: new Date(), data: doc.data() };
      });
    });
}
```

This is something I'll be exploring soon in my projects again with pullstate. So I'm glad you brought it up! Interested to see how we can make them play nicely together.

**Comment by davej** [comment]
Thanks Paul, I'm using Firestore. Thanks for those tips. Yes, reactions are the piece I was missing here I think. It will make things a little cleaner I think and it will keep the pullstate -> firestore logic out of my components.

I've worked on a few open source projects myself and I know it's fun to see how these things get used in the wild so here a quick update on ToDesktop 😉. I [launched the project just under 4 weeks ago on Product Hunt](https://www.producthunt.com/posts/todesktop) and it received a pretty good reception. I have a few customers and a bit of revenue so I'm throwing myself into this full-time.

**Comment by lostpebble** [comment]
That's seriously awesome! Looks like you got some great exposure on Product Hunt, congrats on the successful release and actually starting to generate revenue with your project! That's a great acheivement man. Very cool you get to go full-time into it now.

Maybe there should be something in the Readme.md ? like a "Who's using pullstate" section. Think its awesome that its being used in such a cool project :) Let me know if you'd be keen.

**Comment by davej** [comment]
> Maybe there should be something in the Readme.md ?

Yeah, I would definitely be up for that. If you want to put Pullstate on Stackshare then I would love to add it there as well: https://stackshare.io/todesktop/frontend

**Comment by lostpebble** [comment]
Great!

I'm trying to add Pullstate to Stackshare at the moment, but its telling me that the name is already taken 🤔  - which is weird because when I search for it nothing comes up.

Might just make it "Pullstate JS" then - rather annoying though.

**Comment by davej** [comment]
Oh no, that might be my fault. I added Pullstate to me stack then it prompted me to 'create new', I clicked the button and then backed out at the next step. I've deleted it from my stack now so perhaps it will be available to register. If not, sorry 😕.

**Comment by lostpebble** [comment]
Ah, okay - no worries :)

I'll point the Stackshare team to your post here so they can see and hopefully rename it.

**Comment by jdorfman** [comment]
@davej thanks for adding it for @lostpebble 🎉

@lostpebble I approved it for you: https://stackshare.io/pullstate 😎

Should be in the next newsletter so if you haven't subscribed... https://stackshare.io/weekly 🔥

**Comment by lostpebble** [comment]
Much thanks @jdorfman !

**Comment by lostpebble** [comment]
Hi @davej !

Just dropping by to let you know I've been working on a documentaion site for Pullstate to get ready for a major 1.0.0 release. Finalizing the API and hoping for it to stay rather solid going forward now.

The section which might interest you concerning Firestore is here: https://lostpebble.github.io/pullstate/docs/subscribe

The API has changed slightly. And there are two store "listening" functions available to you depending on the use case, in your case I think a subscription is fine as you only want to watch some state and react by changing the Firebase database listener (which in turn updates your stores). Reactions I am reserving for cases where you actually want to update your store based on other values in your store updating - it batches all the updates together before re-rendering your components.

I actually used your use-case as an example further down 😃 

On that note, is it working out for you? I haven't actually tested the code, its more of a "should work" kinda thing. I'd be interested to know how you solved it at the end.

Anyway, hope things are continuing on the up and up on your side!

**Comment by davej** [comment]
Thanks Paul, docs look great. Excited for 1.0 let me know if you are publishing it to Hacker News or Reddit so that I can lend my support.

I'll update you during the week how Pullstate + Firebase is going. :)

**Comment by lostpebble** [comment]
Thanks Dave, will be sure to let you know when I'm ready for release. Just busy ironing out some final things and making sure the API is completely stable and consistent (as much as possible).

**Comment by davej** [comment]
Awesome, yes the API is working really well. Love it! 😊

When using Subscriptions, is it possible to get the previous value of a watched object? My use case is that I only want to update firebase with the changed fields. So, I want to run the object through a delta function first before sending the updated field to Firebase.

**Comment by lostpebble** [comment]
Ah, interesting case. I've added it to the next version and updated the documentation :)

You can start using it now on `1.0.0-beta.3`

**Comment by davej** [comment]
Thanks Paul!

I think there's a small issue with it though because in the [line before](https://github.com/lostpebble/pullstate/commit/4047a0ca573b47ee68c35906f31e1c56d4ec4761#diff-d3f9f799a39367aa9b3f0678d1a0699dR34) you are setting `lastWatchState = nextWatchState;`. It's the same on Reactions.

So, `watched` and `prevWatched` are the same for me.

**Comment by lostpebble** [comment]
Oh shucks! Yea I see, I'll edit that now!

**Comment by lostpebble** [comment]
Sorry about that! - fixed in `1.0.0-beta.4`

Added a test-case for it too now. Will be doing that in future - need more tests.

**Comment by davej** [comment]
Thanks Paul, works perfectly. 👍

**Comment by lostpebble** [comment]
Hi @davej !

Just published `1.0.0`, Posted on Reddit to try get some exposure :)

https://www.reddit.com/r/reactjs/comments/bzspsx/pullstate_100_released/

Thanks for being a big part of the journey to where pullstate is today!

**Comment by davej** [comment]
Upvoted! :-)
